README.txt
Overview
========

Placeful Workflow is a Plone product that allows you to define workflow
policies that define content type to workflow mappings that can be applied
in any sub-folder of your Plone site:

1. When you access the root of your site, you will see a new action in
   the workflow state drop-down menu called "policy". Click on the 
   "policy" link. 

2. The next page will let you add a policy to your folder by clicking
   on the "Add Workflow policy" link. Click on "Add Workflow policy".

3. Now you have a workflow policy in your site, and you can set the
   workflow policies for this folder and below.

We didn't add any workflow policies, so you don't have a choice
of different workflow policies yet, so the default workflow
policy will be taken both for the folder and below.


Activation
----------

* Use Site Setup -> Add-on Products panel to install

* Now you can define and apply local workflow policies through the Plone
  Interface.

Usage
-----

Now, let's define a new workflow policy:

1. Access "Site Setup" and click on "Placeful Workflow" in the "Add-on
   Product Configuration" section.

2. Enter the name "my_policy" in the "New policy" field, and click on
   "add".

3. Now you have a new policy. Enter the title "Example policy" and the
   description "This is an example policy". 

4. Change the workflow for the content type "Folder" from "folder_workflow"
   to plone_workflow", and click on "Save". Now all your content types
   should use the "plone_workflow".

Let's test the new workflow policy for "Folders" at the root
of our site:

1. At the root of the site, select the "Policy" link in the
   workflow state drop-down menu.

2. Select "Example policy" for "In this Folder" and "Below this Folder" and
   click "Save". 

3. Then, let's add a Folder to see whether the new workflow policy is 
   active. Go to the root of your site and select "Folder" from 
   the "Add new item" drop-down list. Enter the id "myfolder", the title
   "My folder" and the description "This is my folder", and click on
   "Save". 

4. Now, when you access the "State" drop-down list, you will see that
   you have the possibility to "submit" the folder. The submit transition
   only exists in the "plone_workflow", and is absent from the 
   "folder_workflow", which demonstrates that the workflow policy we have
   chosen is used for the "Folder" content type.

Let's go one step further and add a new folder inside of "My folder".
After having added the new folder, you should also find the "Submit"
transition available.

Now it would be interesting to change the workflow policy setting in
the Plone site. Let's first change the workflow policy for "Below this
Folder" to "Default Policy". You will find that the second folder does
not more have the "submit" transition.

You can add an additional workflow policy in the first folder, which
assigns the "My policy" for "In this Folder", so the second folder
will once again have the "submit" transition.

Additional tools
----------------

The Placeful Workflow tool (portal_placeful_workflow) is installed by
the installer. It provides a few configuration options so that you use
to create you workflow policies through the ZMI.



CHANGES
trunk (unreleased)
==================

- Declare package dependencies and fixed deprecation warnings for use
  of Globals.
  [hannosch]

- Define wtool inside the prefs_workflow_policy_mapping template.
  [hannosch]

- Copied safeEditProperty from CMFPlone.migration_util to avoid a dependency.
  [hannosch]


1.4.0 (2008-11-05)
==================

- Port evolutions from the 1.3 maintenance branch (old style product).
  [encolpe]

- Fixed bad version in metadata.xml (again)
  [encolpe]

- Removed Favorite content type.
  [hannosch]

- Fixed "ValueError: 'acquisition' is not a workflow ID" in
  prefs_workflow_policy_mapping.  Fix prepared by jhackel.  Fixes
  http://dev.plone.org/plone/ticket/8101
  [maurits]


1.3.2 (2008-06-30)
==================

- Fixed bad metadata.xml.
  [encolpe]

- Fixed incorrect variable name in exception message.
  [davisagli]


Products/CMFPlacefulWorkflow/tests/__init__.py
# -*- coding: utf-8 -*-
## CMFPlacefulWorkflow
## Copyright (C)2005 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""\
Unit test package

To run all tests type 'python runalltests.py' or bin/zopectl test Products/CMFPlacefulWorkflow
"""
__version__ = "$Revision: 36863 $"
# $Source: /cvsroot/ingeniweb/CMFPlacefulWorkflow/tests/__init__.py,v $
# $Id: __init__.py 36863 2007-02-05 21:30:44Z encolpe $
__docformat__ = 'restructuredtext'


Products/CMFPlacefulWorkflow/tests/CMFPlacefulWorkflowTestCase.py
# -*- coding: utf-8 -*-
## CMFPlacefulWorkflow
## Copyright (C)2005 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
CMFPlacefulWorkflow TestCase module
"""
__version__ = "$Revision: 41292 $"
# $Source: /cvsroot/ingeniweb/CMFPlacefulWorkflow/tests/CMFPlacefulWorkflowTestCase.py,v $
# $Id: CMFPlacefulWorkflowTestCase.py 41292 2007-04-29 21:20:27Z optilude $
__docformat__ = 'restructuredtext'

# Zope imports
from Testing import ZopeTestCase
from AccessControl.SecurityManagement import newSecurityManager
from AccessControl.SecurityManagement import noSecurityManager

# Plone imports
from Products.PloneTestCase import PloneTestCase


class CMFPlacefulWorkflowTestCase(PloneTestCase.PloneTestCase):

    # Globals
    portal_name = 'portal'
    portal_owner = 'portal_owner'
    user_name = PloneTestCase.default_user
    user_password = PloneTestCase.default_password

    class Session(dict):
        def set(self, key, value):
            self[key] = value

    def _setup(self):
        PloneTestCase.PloneTestCase._setup(self)
        self.app.REQUEST['SESSION'] = self.Session()

    def beforeTearDown(self):
        # logout
        noSecurityManager()

    def loginAsPortalMember(self):
        '''Use if you need to manipulate site as a member.'''
        self._setupUser()
        self.mtool.createMemberarea(self.user_name)
        member = self.mtool.getMemberById(self.user_name)
        member.setMemberProperties({'fullname': self.user_name.capitalize(),
                                    'email': 'test@example.com', })
        self.login()

    def loginAsPortalOwner(self):
        '''Use if you need to manipulate site as a manager.'''
        uf = self.app.acl_users
        user = uf.getUserById(self.portal_owner).__of__(uf)
        newSecurityManager(None, user)

    def getPermissionsOfRole(self, role):
        perms = self.portal.permissionsOfRole(role)
        return [p['name'] for p in perms if p['selected']]


class CMFPlacefulWorkflowFunctionalTestCase(
    CMFPlacefulWorkflowTestCase, PloneTestCase.FunctionalTestCase):
    pass

# Install CMFPlacefulWorkflow
ZopeTestCase.installProduct('CMFPlacefulWorkflow')

# Setup Plone site
PloneTestCase.setupPloneSite(id='plone', products=[
    'CMFPlacefulWorkflow',
    ])


Products/CMFPlacefulWorkflow/adapter.py
## -*- coding: utf-8 -*-
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

__version__ = "$Revision: $"
# $Source: $
# $Id: $
__docformat__ = 'restructuredtext'

from plone.app.workflow.interfaces import IWorkflowChain
from plone.app.workflow.workflowchain import ToolWorkflowChain

from zope.interface import Interface, implementer
from zope.component import adapter

from Products.CMFPlacefulWorkflow.PlacefulWorkflowTool import WorkflowPolicyConfig_id
from Products.CMFPlacefulWorkflow.interfaces import IPlacefulMarker
from Acquisition import aq_base, aq_parent, aq_inner
from Products.CMFCore.utils import getToolByName

@adapter(Interface, IPlacefulMarker)
@implementer(IWorkflowChain)
def PlacefulWorkflowChain(ob, tool):
    """Monkey-patched by CMFPlacefulWorkflow to look for placeful workflow configurations.

    Goal: find a workflow chain in a policy

    Steps:
    1. ask the object if it contains a policy
    2. if it does, ask him for a chain
    3. if there's no chain for the type the we loop on the parent
    4. if the parent is the portal object or None we stop and ask
       portal_workflow
    """
    if isinstance(ob, basestring):
        # We are not in an object, then we can only get default from
        # portal_workflow
        return ToolWorkflowChain(ob, tool)

    elif hasattr(aq_base(ob), 'getPortalTypeName'):
        portal_type = ob.getPortalTypeName()
    else:
        portal_type = None

    if portal_type is None or ob is None:
        return ()

    # Inspired by implementation in CPSWorkflowTool.py of CPSCore 3.9.0
    # Workflow needs to be determined by true containment not context
    # so we loop over the actual containers
    chain = None
    wfpolicyconfig = None
    current_ob = aq_inner(ob)
    # start_here is used to check 'In policy': We check it only in the
    # first folder
    start_here = True
    portal = aq_base(getToolByName(tool, 'portal_url').getPortalObject())
    while chain is None and current_ob is not None:
        if getattr(aq_base(current_ob), WorkflowPolicyConfig_id, None) is not None:
            wfpolicyconfig = getattr(current_ob, WorkflowPolicyConfig_id)
            chain = wfpolicyconfig.getPlacefulChainFor(portal_type,
                                                       start_here=start_here)
            if chain is not None:
                return chain

        elif aq_base(current_ob) is portal:
            break
        start_here = False
        current_ob = aq_inner(aq_parent(current_ob))

    # fallback on the default mechanism
    return ToolWorkflowChain(ob, tool)


Products/CMFPlacefulWorkflow/__init__.py
# -*- coding: utf-8 -*-
## CMFPlacefulWorkflow
## Copyright (C)2005 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
Initialization
"""
__version__ = "$Revision: 62604 $"
# $Source: /cvsroot/ingeniweb/CMFPlacefulWorkflow/__init__.py,v $
# $Id: __init__.py 62604 2008-04-13 20:50:14Z hannosch $
__docformat__ = 'restructuredtext'

import PlacefulWorkflowTool
import DefaultWorkflowPolicy
import WorkflowPolicyConfig

from AccessControl import ModuleSecurityInfo
from Products.CMFCore import utils, DirectoryView

tools = (PlacefulWorkflowTool.PlacefulWorkflowTool, )

DirectoryView.registerDirectory('skins', globals())

# Initialization method
def initialize(context):
    utils.registerIcon(
        DefaultWorkflowPolicy.DefaultWorkflowPolicyDefinition,
        'images/workflow_policy.gif',
        globals())

    context.registerClass(
        PlacefulWorkflowTool.PlacefulWorkflowTool,
        meta_type="Placeful Workflow Tool",
        constructors=(PlacefulWorkflowTool.addPlacefulWorkflowTool,),
        icon = 'tool.gif')

    context.registerClass(
        WorkflowPolicyConfig.WorkflowPolicyConfig,
        permission='Add Workflow Policy',
        constructors=(WorkflowPolicyConfig.manage_addWorkflowPolicyConfigForm,
                      WorkflowPolicyConfig.manage_addWorkflowPolicyConfig),
        icon='www/WorkflowPolicyConfig_icon.gif',
    )

    utils.ToolInit( 'CMF Placeful Workflow Tool'
                  , tools=tools
                  , icon='tool.gif'
                  ).initialize( context )

ModuleSecurityInfo('Products.CMFPlacefulWorkflow').declarePublic('CMFPlacefulWorkflowMessageFactory')

# Import "CMFPlacefulWorkflowMessageFactory as _" to create messages
from zope.i18nmessageid import MessageFactory
CMFPlacefulWorkflowMessageFactory = MessageFactory('cmfplacefulworkflow')



Products/CMFPlacefulWorkflow/PlacefulWorkflowTool.py
# -*- coding: utf-8 -*-
## CMFPlacefulWorkflow
## Copyright (C)2005 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
PlacefulWorkflowTool main class
"""
__version__ = "$Revision: 62441 $"
# $Source: /cvsroot/ingeniweb/CMFPlacefulWorkflow/PlacefulWorkflowTool.py,v $
# $Id: PlacefulWorkflowTool.py 62441 2008-04-10 20:03:21Z encolpe $
__docformat__ = 'restructuredtext'

from os.path import join as path_join

from Acquisition import aq_base
from AccessControl.requestmethod import postonly
from AccessControl import Unauthorized
from OFS.Folder import Folder
from OFS.ObjectManager import IFAwareObjectManager
from AccessControl import ClassSecurityInfo
from App.class_init import InitializeClass
from App.Common import package_home

from Products.PageTemplates.PageTemplateFile import PageTemplateFile

from zope.interface import implements
from zope.interface import implementedBy

from Products.CMFCore.utils import getToolByName
from Products.CMFCore.utils import UniqueObject
from Products.CMFCore.permissions import ManagePortal
from Products.CMFCore.permissions import View
from Products.CMFCore.utils import registerToolInterface
from Products.CMFCore.utils import _checkPermission

from Products.CMFCore.interfaces import ISiteRoot
from interfaces import IPlacefulWorkflowTool

WorkflowPolicyConfig_id = ".wf_policy_config"
_MARKER = object()

def addPlacefulWorkflowTool(self,REQUEST={}):
    """
    Factory method for the Placeful Workflow Tool
    """
    id='portal_placeful_workflow'
    pwt=PlacefulWorkflowTool()
    self._setObject(id, pwt, set_owner=0)
    getattr(self, id)._post_init()
    if REQUEST:
        return REQUEST.RESPONSE.redirect(self.absolute_url() + '/manage_main')


def safeEditProperty(obj, key, value, data_type='string'):
    """ An add or edit function, surprisingly useful :) """
    if obj.hasProperty(key):
        obj._updateProperty(key, value)
    else:
        obj._setProperty(key, value, data_type)


class PlacefulWorkflowTool(UniqueObject, Folder, IFAwareObjectManager):
    """
    PlacefulWorkflow Tool
    """

    id = 'portal_placeful_workflow'
    meta_type = 'Placeful Workflow Tool'

    implements(IPlacefulWorkflowTool)

    _actions = []

    security = ClassSecurityInfo()


    manage_options=Folder.manage_options

    def __init__(self):
        # Properties to be edited by site manager
        safeEditProperty(self, 'max_chain_length', 1, data_type='int')

    _manage_addWorkflowPolicyForm = PageTemplateFile(path_join('www', 'add_workflow_policy'), globals())

    security.declareProtected( ManagePortal, 'manage_addWorkflowPolicyForm')
    def manage_addWorkflowPolicyForm(self, REQUEST):
        """ Form for adding workflow policies.
        """
        wfpt = []
        for key in _workflow_policy_factories.keys():
            wfpt.append(key)
        wfpt.sort()
        return self._manage_addWorkflowPolicyForm(REQUEST, workflow_policy_types=wfpt)

    security.declareProtected( ManagePortal, 'manage_addWorkflowPolicy')
    def manage_addWorkflowPolicy(self, id,
                                 workflow_policy_type='default_workflow_policy (Simple Policy)',
                                 duplicate_id='empty',
                                 RESPONSE=None,
                                 REQUEST=None):
        """ Adds a workflow policies from the registered types.
        """
        if id in ('empty', 'portal_workflow'):
            raise ValueError, "'%s' is reserved. Please choose another id." % id

        factory = _workflow_policy_factories[workflow_policy_type]
        ob = factory(id)
        self._setObject(id, ob)

        if duplicate_id and duplicate_id != 'empty':
            types_tool = getToolByName(self, 'portal_types')
            new_wp = self.getWorkflowPolicyById(id)

            if duplicate_id == 'portal_workflow':
                wf_tool = getToolByName(self, 'portal_workflow')

                new_wp.setDefaultChain(wf_tool._default_chain)

                for ptype in types_tool.objectIds():
                    chain = wf_tool.getChainForPortalType(ptype, managescreen=True)
                    if chain:
                        new_wp.setChain(ptype, chain)

            else:
                orig_wp = self.getWorkflowPolicyById(duplicate_id)
                new_wp.setDefaultChain(orig_wp.getDefaultChain('Document'))

                for ptype in types_tool.objectIds():
                    chain = orig_wp.getChainFor(ptype, managescreen=True)
                    if chain:
                        new_wp.setChain(ptype, chain)

        if RESPONSE is not None:
            RESPONSE.redirect(self.absolute_url() +
                              '/manage_main?management_view=Contents')
    manage_addWorkflowPolicy = postonly(manage_addWorkflowPolicy)

    def all_meta_types(self):
        return (
            {'name': 'WorkflowPolicy',
             'action': 'manage_addWorkflowPolicyForm',
             'permission': ManagePortal },)

    security.declareProtected( ManagePortal, 'getWorkflowPolicyById')
    def getWorkflowPolicyById(self, wfp_id):
        """ Retrieve a given workflow policy.
        """
        if wfp_id is None:
            return None
        policy = getattr(self.aq_explicit, wfp_id, _MARKER)
        if policy is not _MARKER:
            if getattr(policy, '_isAWorkflowPolicy', 0):
                return policy
        return None

    security.declarePublic('isValidPolicyName')
    def isValidPolicyName(self, policy_id):
        """ Return True if a policy exist
        """
        return self.getWorkflowPolicyById(policy_id) is not None

    security.declareProtected( ManagePortal, 'getWorkflowPolicies')
    def getWorkflowPolicies(self):
        """ Return the list of workflow policies.
        """
        wfps = []
        for obj_name, obj in self.objectItems():
            if getattr(obj, '_isAWorkflowPolicy', 0):
                wfps.append(obj)
        return tuple(wfps)

    security.declarePublic('getWorkflowPolicyIds')
    def getWorkflowPolicyIds(self):
        """ Return the list of workflow policy ids.
        """
        wfp_ids = []

        for obj_id, obj in self.objectItems():
            if getattr(obj, '_isAWorkflowPolicy', 0):
                wfp_ids.append(obj_id)

        return tuple(wfp_ids)

    security.declarePublic('getWorkflowPolicyInfos')
    def getWorkflowPolicyInfos(self):
        """ Return the list of workflow policy ids.
        """
        wfp_ids = []

        for obj_id, obj in self.objectItems():
            if getattr(obj, '_isAWorkflowPolicy', 0):
                wfp_ids.append({'id': obj_id, 'title': obj.title_or_id(),
                                'description': obj.description})

        return tuple(wfp_ids)


    security.declareProtected( View, 'getWorkflowPolicyConfig')
    def getWorkflowPolicyConfig(self, ob):
        """ Return a local workflow configuration if it exist
        """
        if ISiteRoot.providedBy(ob):
            # Site root use portal_workflow tool as local policy
            return None
        if not _checkPermission(ManagePortal, ob):
            raise Unauthorized("You need %s permission on %s" % (
                ManagePortal, '/'.join(ob.getPhysicalPath())))

        return getattr(ob.aq_explicit, WorkflowPolicyConfig_id, None)

    def _post_init(self):
        """
        _post_init(self) => called from manage_add method, acquired within ZODB (__init__ is not)
        """
        pass

    #
    #   portal_workflow_policy implementation.
    #

    def getMaxChainLength(self):
        """Return the max workflow chain length"""
        max_chain_length = self.getProperty('max_chain_length')
        return max_chain_length

    def setMaxChainLength(self, max_chain_length):
        """Set the max workflow chain length"""
        safeEditProperty(self, 'max_chain_length', max_chain_length, data_type='int')

_workflow_policy_factories = {}

def _makeWorkflowPolicyFactoryKey(factory, id=None, title=None):
    # The factory should take one argument, id.
    if id is None:
        id = getattr(factory, 'id', '') or getattr(factory, 'meta_type', '')
    if title is None:
        title = getattr(factory, 'title', '')
    key = id
    if title:
        key = key + ' (%s)' % title
    return key

def addWorkflowPolicyFactory(factory, id=None, title=None):
    key = _makeWorkflowPolicyFactoryKey( factory, id, title )
    _workflow_policy_factories[key] = factory

def _removeWorkflowPolicyFactory( factory, id=None, title=None ):
    """ Make teardown in unitcase cleaner. """
    key = _makeWorkflowPolicyFactoryKey( factory, id, title )
    try:
        del _workflow_policy_factories[key]
    except KeyError:
        pass

InitializeClass(PlacefulWorkflowTool)
registerToolInterface('portal_placeful_workflow', IPlacefulWorkflowTool)


Products/CMFPlacefulWorkflow/DefaultWorkflowPolicy.py
# -*- coding: utf-8 -*-
## CMFPlacefulWorkflow
## Copyright (C)2005 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
A simple workflow policy.
"""
__version__ = "$Revision: 62540 $"
# $Source: /cvsroot/ingeniweb/CMFPlacefulWorkflow/DefaultWorkflowPolicy.py,v $
# $Id: DefaultWorkflowPolicy.py 62540 2008-04-12 07:50:01Z encolpe $
__docformat__ = 'restructuredtext'

from os.path import join as path_join

from zope.interface import implements

from AccessControl import ClassSecurityInfo
from AccessControl.requestmethod import postonly
from App.class_init import InitializeClass
from Persistence import PersistentMapping
from Acquisition import aq_base

from Products.PageTemplates.PageTemplateFile import PageTemplateFile

from Products.CMFCore.utils import SimpleItemWithProperties
from Products.CMFPlacefulWorkflow.PlacefulWorkflowTool import addWorkflowPolicyFactory

from Products.CMFCore.permissions import ManagePortal

from Products.CMFPlacefulWorkflow.interfaces.portal_placeful_workflow import IWorkflowPolicyDefinition
from Products.CMFPlacefulWorkflow.global_symbols import Log, LOG_DEBUG

from Products.CMFCore.utils import getToolByName

DEFAULT_CHAIN = '(Default)'
_MARKER = '_MARKER'

class DefaultWorkflowPolicyDefinition(SimpleItemWithProperties):

    implements(IWorkflowPolicyDefinition)

    meta_type = 'WorkflowPolicy'
    id = 'default_workflow_policy'
    _isAWorkflowPolicy = 1

    _chains_by_type = None  # PersistentMapping
    _default_chain = None # Fallback to wf tool

    security = ClassSecurityInfo()

    manage_options = (
        { 'label' : 'Workflows', 'action' : 'manage_main'},
    )

    #
    #   ZMI methods
    #

    security.declareProtected( ManagePortal, '_manage_workflows' )
    _manage_workflows = PageTemplateFile(path_join('www', 'define_local_workflow_policy'),
                                              globals(),
                                              __name__= 'manage_main')

    def __init__(self, id):
        self.id = id
        self.title = ''
        self.description = ''

    security.declareProtected( ManagePortal, 'getId')
    def getId(self):
        """ Return the id
        """
        return self.id

    security.declareProtected( ManagePortal, 'getTitle')
    def getTitle(self):
        """ Return the title
        """
        title = getattr(self, 'title', '')
        return title

    security.declareProtected( ManagePortal, 'getDescription')
    def getDescription(self):
        """ Return the description
        """
        description = getattr(self, 'description', '')
        return description

    security.declareProtected( ManagePortal, 'setTitle')
    def setTitle(self, title):
        """ Set the title
        """
        self.title=title

    security.declareProtected( ManagePortal, 'setDescription')
    def setDescription(self, description):
        """ Set the description
        """
        self.description = description

    security.declareProtected( ManagePortal, 'manage_main')
    def manage_main(self, REQUEST, manage_tabs_message=None):
        """ Show a management screen for changing type to workflow connections

        Display 'None' if there's no chain for a type.
        """
        cbt = self._chains_by_type
        ti = self._listTypeInfo()
        types_info = []
        for t in ti:
            id = t.getId()
            title = t.Title()
            if title == id:
                title = None

            if cbt is not None and cbt.has_key(id):
                chain = ', '.join(cbt[id])
            else:
                chain = 'None'

            types_info.append({
                'id': id,
                'title': title,
                'chain': chain,
                #'cbt': repr(cbt.get(id)), # for debug purpose
            })
        return self._manage_workflows(
            REQUEST,
            default_chain=', '.join(self._default_chain or ()),
            types_info=types_info,
            management_view='Workflows',
            manage_tabs_message=manage_tabs_message)


    security.declareProtected( ManagePortal, 'manage_changeWorkflows')
    def manage_changeWorkflows(self, title, description, default_chain, props=None, REQUEST=None):
        """ Changes which workflows apply to objects of which type

        A chain equal to 'None' is empty we remove the entry.
        """
        self.title = title
        self.description = description

        wf_tool = getToolByName(self, 'portal_workflow')

        if props is None:
            props = REQUEST
        cbt = self._chains_by_type
        if cbt is None:
            self._chains_by_type = cbt = PersistentMapping()
        ti = self._listTypeInfo()
        # Set up the chains by type.
        for t in ti:
            id = t.getId()
            field_name = 'chain_%s' % id
            chain = props.get(field_name, DEFAULT_CHAIN).strip()

            if chain == 'None':
                if cbt.get(id, _MARKER) is not _MARKER:
                    self.delChain(id)
                continue

            self.setChain(id, chain)

        # Set up the default chain.
        self.setDefaultChain(default_chain)
        if REQUEST is not None:
            return self.manage_main(REQUEST, manage_tabs_message='Changed.')
    manage_changeWorkflows = postonly(manage_changeWorkflows)

    security.declareProtected( ManagePortal, 'setChainForPortalTypes')
    def setChainForPortalTypes(self, pt_names, chain, REQUEST=None):
        """ Set a chain for portal types.
        """
        for portal_type in pt_names:
            self.setChain(portal_type, chain)
    setChainForPortalTypes = postonly(setChainForPortalTypes)

    security.declareProtected( ManagePortal, 'getChainFor')
    def getChainFor(self, ob, managescreen=False):
        """Returns the chain that applies to the object.

        If chain doesn't exist we return None to get a fallback from
        portal_workflow.

        @parm managescreen: If True return the special tuple
                            ('default') instead of the actual default
                            chain (hack).
        """

        cbt = self._chains_by_type
        if type(ob) == type(''):
            pt = ob
        elif hasattr(aq_base(ob), '_getPortalTypeName'):
            pt = ob._getPortalTypeName()
        else:
            pt = None

        if pt is None:
            return None

        chain = None
        if cbt is not None:
            chain = cbt.get(pt, _MARKER)

        # Backwards compatibility: before chain was a string, not a list
        if chain is not _MARKER and type(chain) == type(''):
            chain = map( lambda x: x.strip(), chain.split(',') )

        Log(LOG_DEBUG, 'Chain founded in policy', chain)
        if chain is _MARKER or chain is None:
            return None
        elif len(chain) == 1 and chain[0] == DEFAULT_CHAIN:
            default = self.getDefaultChain(ob)
            if default:
                if managescreen:
                    return chain[0]
                else:
                    return default
            else:
                return None

        return chain

    security.declareProtected( ManagePortal, 'setDefaultChain')
    def setDefaultChain(self, default_chain, REQUEST=None):

        """ Sets the default chain for this tool. """
        wftool = getToolByName(self, 'portal_workflow')

        if type(default_chain) is type(''):
            default_chain = map( lambda x: x.strip(), default_chain.split(',') )
        ids = []
        for wf_id in default_chain:
            if wf_id:
                if not wftool.getWorkflowById(wf_id):
                    raise ValueError, ("'%s' is not a workflow ID.\nchain: %s" % (
                        wf_id, repr(default_chain)))
                ids.append(wf_id)

        self._default_chain = tuple(ids)
    setDefaultChain = postonly(setDefaultChain)

    security.declareProtected( ManagePortal, 'getDefaultChain')
    def getDefaultChain(self, ob):
        """ Returns the default chain."""
        if self._default_chain is None:
            wf_tool = getToolByName(self, 'portal_workflow')
            return wf_tool.getDefaultChainFor(ob)
        else:
            return self._default_chain

    security.declareProtected( ManagePortal, 'setChain')
    def setChain(self, portal_type, chain, REQUEST=None):
        """Set the chain for a portal type. 

           @type chain: tuple of strings or None 
           @param chain: A tuple of workflow ids to be set for the portal type. 
                         A few special values exsist:  
                           - C{None}: Acquire chain from a policy above,  
                                      ultimatly from the portal workflow settings. 
                           - C{()} (empty tuple): No workflow for this type. 
                           - C{('default',)}: Use the configured default workflow. 
        """ 
        # Verify input data
        if portal_type not in [pt.id for pt in self._listTypeInfo()]:
            raise ValueError, ("'%s' is not a valid portal type." % portal_type)

        if type(chain) is type(''):
            chain = map( lambda x: x.strip(), chain.split(',') )

        wftool = getToolByName(self, 'portal_workflow')
        cbt = self._chains_by_type
        if cbt is None:
            self._chains_by_type = cbt = PersistentMapping()

        # if chain is None or default, we remove the entry
        if chain is None:
            if cbt.has_key(portal_type):
                del cbt[portal_type]
        elif len(chain) == 1 and chain[0] == DEFAULT_CHAIN:
            cbt[portal_type] = chain
        else:
            for wf_id in chain:
                if wf_id != '' and not wftool.getWorkflowById(wf_id):
                    raise ValueError, ("'%s' is not a workflow ID.\nchain: %s" % (
                        wf_id, repr(chain)))
            cbt[portal_type] = tuple(chain)
    setChain = postonly(setChain)

    security.declareProtected( ManagePortal, 'delChain')
    def delChain(self, portal_type, REQUEST=None):
        """Delete the chain for a portal type."""
        if self._chains_by_type.has_key(portal_type):
            del self._chains_by_type[portal_type]
    delChain = postonly(delChain)

    #
    #   Helper methods
    #
    security.declarePrivate( '_listTypeInfo' )
    def _listTypeInfo(self):

        """ List the portal types which are available.
        """
        pt = getToolByName(self, 'portal_types', None)
        if pt is None:
            return ()
        else:
            return pt.listTypeInfo()


InitializeClass(DefaultWorkflowPolicyDefinition)

addWorkflowPolicyFactory(DefaultWorkflowPolicyDefinition, title='Simple Policy')

